ext/search/Civi/Api4/Action/SearchDisplay/Get.php

```diff
@@ -0,0 +1,66 @@
+<?php
+
+/*
+ +--------------------------------------------------------------------+
+ | Copyright CiviCRM LLC. All rights reserved.                        |
+ |                                                                    |
+ | This work is published under the GNU AGPLv3 license with some      |
+ | permitted exceptions and without any warranty. For full license    |
+ | and copyright information, see https://civicrm.org/licensing       |
+ +--------------------------------------------------------------------+
+ */
+
+/**
+ *
+ * @package CRM
+ * @copyright CiviCRM LLC https://civicrm.org/licensing
+ */
+
+namespace Civi\Api4\Action\SearchDisplay;
+
+use Civi\Api4\Generic\Result;
+
+/**
+ * @inheritDoc
+ */
+class Get extends \Civi\Api4\Generic\DAOGetAction {
+
+  /**
+   * @param \Civi\Api4\Generic\Result $result
+   */
+  protected function getObjects(Result $result) {
+    parent::getObjects($result);
+
+    // Check permissions before presenting fields as editable
+    if ($this->checkPermissions) {
+      foreach ($result as $idx => $display) {
+        foreach ($display['settings']['columns'] ?? [] as $order => $column) {
+          if (!empty($column['editable'])) {
+            if (!$this->checkEditPermission($column['editable']['entity'])) {
+              unset($result[$idx]['settings']['columns'][$order]['editable']);
+            }
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * @param string $entity
+   * @return bool
+   */
+  private function checkEditPermission(string $entity): bool {
+    switch ($entity) {
+      // FIXME: This errs on the side of being overly restrictive for contacts, as the user may not have "edit all contacts"
+      // permission, but be granted edit rights to a set of contacts via ACLs
+      case 'Contact':
+        return \CRM_Core_Permission::check('edit all contacts');
+
+      // FIXME: Other entities have the opposite problem, and may be under-restrictive, as the user may have permission
```

Comment:
@colemanw so I'm not fully following here - can we flesh this comment out a bit more with an example / why the option chosen addresses it- obviously we want to err on the side of being restrictive

Reply:
@eileenmcnaughton basically the `'default':` case here is checking the gate-keeper permission for `api.update`. For example, the gate-keeper permission for `Campaign::update` is `('administer CiviCampaign' OR 'manage campaign')`.

But the `Contact::update` has no gate-keeper permission, because all actions on that entity are managed by ACLs. That's why I added a separate case for it. (and FYI no this new edit-in-place feature is not limited to contacts, it's limited to non-readonly fields, so you can't e.g. edit an `id` or a contact's `display_name`)

The 'FIXME' comments are to say that the solution is imperfect. And honestly I'm not sure I agree we ought to err on the side of restrictive. If we are wrong, then the result is users getting locked-out from something they ought to be allowed to do. OTOH if we are under-restrictive, then the worst that happens is an error message when the user tries to edit something and can't.

The more I think about this, I want to just delete this commit from the PR and rework it separately. I'd like to see this PR merged so it can be beta-tested and not blocked by presentation-layer permissions. I'd rather work up a solution that checks fine-grained ACLs per row, and applies them not only to the edit-in-place fields but also to the links and buttons and menus that might be linking to a screen the user doesn't have permission to see (e.g. Delete Contact).

Reply:
@colemanw I guess you are right that it's a usability issue not a security bypass - I'm OK to merge this as is if you want to keep working on it.

I was wanting to see more comments in the actual code - but of course it might be changing again

I'm a bit unsure if we have gone the wrong track not having equivelents to 'edit all contacts' like 'edit all contributions'

Reply:
My thought is to rip out that last commit which checks coarse-grained permissions and instead create a new api action:

```php
// Check if the current user has permission to update participant id 123
civicrm_api4('Participant', 'check', [
  'user_id' => 'current_user', // this would be the default
  'action' => 'update',
  'where' => [['id', '=', 123]], // could potentially check multiple records at once
);
```

This action would check coarse-grained permissions first, and then check ACLs if the entity has any.

This api would give us a consistent way to check permissions for any entity/action combo without actually doing the action (I guess that's pretty similar to @totten's `dry_run` concept).